_format/1.1/index.md

```diff
@@ -1970,6 +1969,19 @@ server, and we are defining its semantics for JSON:API.
 
 #### <a href="#crud-updating-relationship-responses" id="crud-updating-relationship-responses" class="headerlink"></a> Responses
 
+##### <a href="#crud-updating-relationship-responses-200" id="crud-updating-relationship-responses-200" class="headerlink"></a> 200 OK
+
+If a server accepts an update but also changes the targeted relationship(s)
+in other ways than those specified by the request, it **MUST** return a `200
+OK` response and a document that includes the updated relationship(s) in its
+primary data.
```

Comment:
What does "include the updated relationship(s)" mean? How does an example response payload look like?

Assuming we have a relationship `tags` of an `articles` resource. Let's assume that currently two resources are referenced by that relationship: `{ "tags": "1", "tags": 2 }`

A client sends a request to add `{ "tags": "3" }` additionally. Due to some domain-specific constraints the same article can not be related to `{ "tags": "3" }` and `{ "tags": "1" }`. The server removes `{ "tags": "1" }` from the relationship automatically.

What does a response payload look like in this case?

The wording above (and in the specification right now) could be interpreted in two ways:

1. Response with all resources currently associated to that relationship.
2. Response with the resources, which are added or removed from that relationship due to the update.

The first one would be easy. The server would response with:

```json
{
  "data": [
    { "tags": "2" },
    { "tags": "3" }
  ]
}
```

The second one is not supported by the base spec. As far as I know the server can not inform the client that `{ "tags": "1" }` has been removed from the relationship.

Not sure if you want to address this one in this pull request as well. Or keep it for a follow-up PR.

Reply:
As the maintainer of [JsonApiDotNetCore](https://github.com/json-api-dotnet/JsonApiDotNetCore), the leading server-side JSON:API implementation in ASP.NET, I can say that we're never going to implement option 1. Because its potentially very expensive to determine and useless to clients, except for theoretical cases. Consider an audit log, where old entries are purged as a side-effect. So POSTing to this to-many relationship would require returning thousands of entries in the response (there's no pagination). Option 2 is so extremely unlikely to ever occur that we can't justify doing extra queries to detect this.

I can see the elegance in using similar statements throughout the spec, but it's pretty impractical when it comes to relationships. Another case is using rowversion/xmin/timestamp columns for optimistic concurrency. They change all the time, but depending on at which side the foreign key is defined, we cannot consistently return that information, so honestly I don't see the point.

From a spec perspective, I think the only sensible solution is to change MUST to MAY in:
> If a server accepts an update but also changes the targeted relationship in other ways than those specified by the request, it MUST return a 200 OK response and a document that includes the updated relationship data as its primary data.

Reply:
> Because it's (...) useless to clients, except for theoretical cases.

I disagree with this. It eases handling of client-side caches as the client can be sure that local cache is still up-to-date. Otherwise a client either needs to refetch the data or hope that no side-effects happened to manage client-side cache.

> Another case is using rowversion/xmin/timestamp columns for optimistic concurrency. They change all the time, but depending on at which side the foreign key is defined, we cannot consistently return that information, so honestly I don't see the point.

The specification is only talking about changes to the relationship in the sense of additional resources being added to or removed from that relationship than those explicitly requested by the client. Additional meta data a server may store with a relationship (or even attributes of a related resource) are not relevant to determine if "the targeted relationship (changes) in other ways than those specified by the request". At least in my reading of the specification.

I see your point though that for some edge cases returning all resource identifier objects for related resources may be too expensive. This may be the case for 1) to-many relationships, 2) having many related resources, which cannot be queried in an efficient way, and 3) adding or removing a related resource causes side-effects on other resources being added.

Reply:
If the goal is to handle client-side caches, then the response structure should be much more flexible. It should be possible to return type/id entries of any resource type, as well as full resource objects of any type. The choice to _only_ return type/id entries that occurred in the request seems arbitrary and not very practical.

Please prove me wrong and show that it's not useless, by providing a real-world compelling example where this makes sense. I haven't found any.